histogram/src/sliding_window.rs

```diff
@@ -0,0 +1,470 @@
+//! A histogram that stores a distribution for a fixed window of time.
+
+use crate::{
+    AtomicInstant, BuildError, Config, Duration, Error, Instant, Ordering, Range, Snapshot,
+    UnixInstant,
+};
+use core::sync::atomic::AtomicU64;
+
+/// A type of histogram that reports on the distribution of values across a
+/// moving window of time. For example, the distribution of values for the past
+/// minute. Internally, this uses atomic counters to allow concurrent
+/// modification.
+pub struct Histogram {
+    config: Config,
+    interval: Duration,
+    span: Duration,
+    started: UnixInstant,
+    tick_origin: Instant,
+    tick_at: AtomicInstant,
+    num_slices: usize,
+    snapshots: Box<[Box<[AtomicU64]>]>,
```

Comment:
Is it necessary to have the snapshots as atomic histograms, given we only ever update the (one) live slice?

Reply:
Unless we want to do an Box<[Arc<Mutex<Box<[u64]>>>]> and take a lock to update the slice? We need some kind of interior mutability so that all operations can take an immutable &self.

Reply:
Alternatively, we can wrap the snapshots in an Arc<Mutex<..>> and take a single lock. Consequence there is that we can't then snapshot in parallel (unclear if that's a pro or con).

Reply:
Resolved by dropping the ring buffer